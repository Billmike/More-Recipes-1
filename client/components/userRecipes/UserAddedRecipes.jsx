import React from 'react';
import PropTypes from 'prop-types';

import UserRecipeCard from './UserRecipeCard';
import Spinner from '../common/Spinner';

const propTypes = {
  fetchUserRecipes: PropTypes.func.isRequired,
  currentProfileUserId: PropTypes.number.isRequired,
  userRecipes: PropTypes.shape({
    isFetchingUserRecipes: PropTypes.bool,
    userAddedRecipes: PropTypes.arrayOf(PropTypes.shape()),
  }).isRequired,
};

/**
 * Class representing user added recipes
 *
 * @class UserAddedRecipes
 *
 * @extends {React.Component}
 */
class UserAddedRecipes extends React.Component {
  /**
   * Component did mount life cycle method
   *
   * @returns {any} Fetches user added recipes
   *
   * @memberof UserAddedRecipes
   */
  componentDidMount() {
    const { fetchUserRecipes, currentProfileUserId } = this.props;

    fetchUserRecipes(currentProfileUserId);
  }

  /**
   * Render user recipes
   *
   * @param {object} props - Properties passed to the function
   *
   * @returns {object} React element
   *
   * @memberof UserAddedRecipes
   */
  renderUserRecipes = (props) => {
    const { userRecipes: { userAddedRecipes } } = props;

    return (
      <div>
        {userAddedRecipes &&
          userAddedRecipes.length === 0
          ? <div className="center-align" >
            <h5>You have not added any recipe</h5>
            <i className="material-icons large">folder_open</i>
          </div>
          : userAddedRecipes.map(recipe => (
            <UserRecipeCard key={recipe.id} recipe={recipe} {...this.props} />
          ))
        }
      </div>
    );
  }

  /**
   * Render method
   *
   * @returns {object} React element
   *
   * @memberof UserAddedRecipes
   */
  render() {
    const { userRecipes: { isFetchingUserRecipes } } = this.props;

    return (
      <div className="row">
        {isFetchingUserRecipes
          ? <div className="center-align">
            <Spinner size="big" />
          </div>
          : this.renderUserRecipes(this.props)
        }
      </div>
    );
  }
}

UserAddedRecipes.propTypes = propTypes;

export default UserAddedRecipes;
