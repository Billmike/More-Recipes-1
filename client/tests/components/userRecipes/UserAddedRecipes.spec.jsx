import React from 'react';

import UserAddedRecipes from '../../../components/userRecipes/UserAddedRecipes';

describe('UserAddedRecipes component', () => {
  it('should render component', () => {
    const props = {
      fetchUserRecipes: jest.fn(),
      currentProfileUserId: 1,
      userRecipes: {
        isFetchingUserRecipes: false,
        userAddedRecipes: [],
      },
    };
    const wrapper = shallow(<UserAddedRecipes {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
  it('should render spinner', () => {
    const props = {
      fetchUserRecipes: jest.fn(),
      currentProfileUserId: 1,
      userRecipes: {
        isFetchingUserRecipes: true,
        userAddedRecipes: [],
      },
    };
    const wrapper = shallow(<UserAddedRecipes {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper.find('Spinner').length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });
  it('should render recipes fetched', () => {
    const props = {
      fetchUserRecipes: jest.fn(),
      currentProfileUserId: 1,
      userRecipes: {
        isFetchingUserRecipes: false,
        userAddedRecipes: [{
          title: 'Beans',
          category: 'Main',
          description: 'Tasty beans',
          preparationTime: 85,
          ingredients: 'Beans, onions, palm oil, salt',
          directions: 'Do this. Do that',
        }, {
          title: 'Beans and rice',
          category: 'Dinner',
          description: 'Tasty beans and rice',
          preparationTime: 95,
          ingredients: 'Beans, rice, onions, palm oil, salt',
          directions: 'Do this. Do this. Do that',
        }],
      },
    };
    const wrapper = shallow(<UserAddedRecipes {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper.find('UserRecipeCard').length).toEqual(2);
    expect(wrapper).toMatchSnapshot();
  });
});
