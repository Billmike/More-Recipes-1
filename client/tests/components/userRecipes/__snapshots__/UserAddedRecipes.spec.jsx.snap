// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserAddedRecipes component should render component 1`] = `
<div
  className="row"
>
  <div>
    <div
      className="center-align"
    >
      <h5>
        No recipes yet
      </h5>
      <i
        className="material-icons large"
      >
        folder_open
      </i>
    </div>
  </div>
</div>
`;

exports[`UserAddedRecipes component should render recipes fetched 1`] = `
<div
  className="row"
>
  <div>
    <UserRecipeCard
      authenticatedUserId={null}
      currentProfileUserId={1}
      fetchUserRecipes={
        [MockFunction] {
          "calls": Array [
            Array [
              1,
            ],
          ],
        }
      }
      recipe={
        Object {
          "category": "Main",
          "description": "Tasty beans",
          "directions": "Do this. Do that",
          "ingredients": "Beans, onions, palm oil, salt",
          "preparationTime": 85,
          "title": "Beans",
        }
      }
      userRecipes={
        Object {
          "isFetchingUserRecipes": false,
          "userAddedRecipes": Array [
            Object {
              "category": "Main",
              "description": "Tasty beans",
              "directions": "Do this. Do that",
              "ingredients": "Beans, onions, palm oil, salt",
              "preparationTime": 85,
              "title": "Beans",
            },
            Object {
              "category": "Dinner",
              "description": "Tasty beans and rice",
              "directions": "Do this. Do this. Do that",
              "ingredients": "Beans, rice, onions, palm oil, salt",
              "preparationTime": 95,
              "title": "Beans and rice",
            },
          ],
        }
      }
    />
    <UserRecipeCard
      authenticatedUserId={null}
      currentProfileUserId={1}
      fetchUserRecipes={
        [MockFunction] {
          "calls": Array [
            Array [
              1,
            ],
          ],
        }
      }
      recipe={
        Object {
          "category": "Dinner",
          "description": "Tasty beans and rice",
          "directions": "Do this. Do this. Do that",
          "ingredients": "Beans, rice, onions, palm oil, salt",
          "preparationTime": 95,
          "title": "Beans and rice",
        }
      }
      userRecipes={
        Object {
          "isFetchingUserRecipes": false,
          "userAddedRecipes": Array [
            Object {
              "category": "Main",
              "description": "Tasty beans",
              "directions": "Do this. Do that",
              "ingredients": "Beans, onions, palm oil, salt",
              "preparationTime": 85,
              "title": "Beans",
            },
            Object {
              "category": "Dinner",
              "description": "Tasty beans and rice",
              "directions": "Do this. Do this. Do that",
              "ingredients": "Beans, rice, onions, palm oil, salt",
              "preparationTime": 95,
              "title": "Beans and rice",
            },
          ],
        }
      }
    />
  </div>
</div>
`;

exports[`UserAddedRecipes component should render spinner 1`] = `
<div
  className="row"
>
  <div
    className="center-align"
  >
    <Spinner
      size="big"
    />
  </div>
</div>
`;
